x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -2003,6 +2006,30 @@ private void reloadRemoteClusterCredentials(Settings settingsWithKeystore) {
         future.actionGet();
     }
 
+    public Map<String, String> getAuthContextForSlowLog() {
+        if (this.securityContext.get() != null && this.securityContext.get().getAuthentication() != null) {
+            Authentication authentication = this.securityContext.get().getAuthentication();
+            Subject authenticatingSubject = authentication.getAuthenticatingSubject();
+            Subject effetctiveSubject = authentication.getEffectiveSubject();
+            Map<String, String> authContext = new HashMap<>();
+            if (authenticatingSubject.getUser() != null) {
+                authContext.put("user.name", authenticatingSubject.getUser().principal());
+                authContext.put("user.realm", authenticatingSubject.getRealm().getName());
+            }
+            if (effetctiveSubject.getUser() != null) {
+                authContext.put("user.effective.name", effetctiveSubject.getUser().principal());
+                authContext.put("user.effective.realm", effetctiveSubject.getRealm().getName());
+            }
+            authContext.put("auth.type", authentication.getAuthenticationType().name());
+            if (authentication.isApiKey()) {
+                authContext.put("apikey.id", authenticatingSubject.getMetadata().get(AuthenticationField.API_KEY_ID_KEY).toString());
+                authContext.put("apikey.name", authenticatingSubject.getMetadata().get(AuthenticationField.API_KEY_NAME_KEY).toString());
+            }
+            return authContext;
+        }
```

Comment:
@jakelandis what do you think about adding the token name here if it's a service account `getAuthenticatingSubject().getMetadata().get(ServiceAccountSettings.TOKEN_NAME_FIELD)`? Would that be helpful for debugging?

Reply:
Good question, but I don't think so.  If I am reading the code correctly the principal will be `namespace/service` i.e. `elastic/kibana` which for this purpose should be sufficient to know who is generating the request. 

The name field can be used to know the exact token that was used to authenticate (the same service account can have multiple credentials that point back to the same service account) but I am not sure if that would be useful in this context. Additionally, I am not sure how often multiple credentials for the same service account are used (It is used for Fleet, but generally not for Kibana (but it can be)). Since this value is most likely omitted as it is optional in the request it most likely a random value and matching a random value back to culprit isn't very friendly nor is using the credential to hunt down the exact client very effecient. To hunt down the exact client, I would think x-opaque-id or matching it to an ip address from other logs would probably be best. Since all service accounts are owned/managed by Elastic if there is a problem we are the ones to fix and just knowing the log info + service is likely sufficient for us to find and fix.